File: futureagi/tracer/services/clickhouse/v2/property_catalog/installation_identity.py
"""Durable, shared installation identity; never an operator version counter.

The lifecycle process is the sole initializer. The sequencer reads the same
file from its persistent fence volume. Database and Kafka destinations are
bound into the identity so an upgrade cannot accidentally reuse another
catalog's spool. Choosing/adopting an identity requires database evidence at
the call site; absence of this file is NOT evidence of an empty database.
"""

from __future__ import annotations

import hashlib
import json
import os
import re
import stat
import tempfile
from collections.abc import Callable
from dataclasses import asdict, dataclass, fields
from pathlib import Path
from uuid import UUID, uuid4

from .mutation_lock import FileCatalogMutationSerializer

IDENTITY_FILENAME = "runtime-identity-v1.json"
IDENTITY_FORMAT = "futureagi.property-catalog-runtime-identity"
IDENTITY_VERSION = 1
PROJECTION_FORMAT = "futureagi.property-catalog-values.v1"
INITIAL_CATALOG_EPOCH = 1
INITIAL_PROJECTION_VERSION = 1
MAX_IDENTITY_BYTES = 4096
_DATABASE = re.compile(r"[a-z][a-z0-9_]{0,127}\Z")
_TOPIC = re.compile(r"[A-Za-z0-9._-]{1,249}\Z")


class InstallationIdentityError(ValueError):
    """Missing, conflicting, unsupported, or corrupt installation metadata."""


def _canonical(value: dict) -> bytes:
    return json.dumps(
        value, ensure_ascii=True, sort_keys=True, separators=(",", ":")
    ).encode("ascii")


@dataclass(frozen=True, slots=True)
class InstallationIdentity:
    environment: str
    target_database: str
    candidate_topic: str
    ordered_topic: str
    catalog_epoch: int
    projection_version: int
    producer_stream_id: str
    projection_format: str = PROJECTION_FORMAT

    def __post_init__(self) -> None:
        if self.environment not in {"development", "production"}:
            raise InstallationIdentityError("identity environment is unsupported")
        if not isinstance(self.target_database, str) or not _DATABASE.fullmatch(
            self.target_database
        ):
            raise InstallationIdentityError("identity target database is invalid")
        if self.target_database in {"default", "system", "information_schema"}:
            raise InstallationIdentityError(
                "identity target is not an isolated catalog"
            )
        for topic in (self.candidate_topic, self.ordered_topic):
            if (
                not isinstance(topic, str)
                or not _TOPIC.fullmatch(topic)
                or topic in {".", ".."}
            ):
                raise InstallationIdentityError("identity Kafka topic is invalid")
        if self.candidate_topic == self.ordered_topic:
            raise InstallationIdentityError("identity requires distinct Kafka topics")
        for name in ("catalog_epoch", "projection_version"):
            value = getattr(self, name)
            if type(value) is not int or not 0 < value < 65536:
                raise InstallationIdentityError(
                    f"identity {name} must be a positive UInt16"
                )
        if self.projection_format != PROJECTION_FORMAT:
            raise InstallationIdentityError("identity projection format is unsupported")
        try:
            parsed = UUID(self.producer_stream_id)
        except (AttributeError, TypeError, ValueError) as exc:
            raise InstallationIdentityError(
                "identity producer stream must be a UUID"
            ) from exc
        if str(parsed) != self.producer_stream_id or not parsed.int:
            raise InstallationIdentityError(
                "identity producer stream must be a canonical nonzero UUID"
            )

    @classmethod
    def fresh(
        cls,
        *,
        environment: str,
        target_database: str,
        candidate_topic: str,
        ordered_topic: str,
    ) -> InstallationIdentity:
        """Call only after proving an empty, isolated destination and spool."""
        return cls(
            environment=environment,
            target_database=target_database,
            candidate_topic=candidate_topic,
            ordered_topic=ordered_topic,
            catalog_epoch=INITIAL_CATALOG_EPOCH,
            projection_version=INITIAL_PROJECTION_VERSION,
            producer_stream_id=str(uuid4()),
        )

    def encode(self) -> bytes:
        unsigned = {
            "format": IDENTITY_FORMAT,
            "version": IDENTITY_VERSION,
            **asdict(self),
        }
        digest = hashlib.sha256(_canonical(unsigned)).hexdigest()
        return _canonical({**unsigned, "identity_sha256": digest}) + b"\n"

    @classmethod
    def decode(cls, raw: bytes) -> InstallationIdentity:
        if len(raw) > MAX_IDENTITY_BYTES or not raw.endswith(b"\n"):
            raise InstallationIdentityError(
                "identity must be a bounded canonical JSON line"
            )
        try:
            document = json.loads(raw)
        except (ValueError, UnicodeError) as exc:
            raise InstallationIdentityError("identity JSON is invalid") from exc
        if not isinstance(document, dict) or _canonical(document) + b"\n" != raw:
            raise InstallationIdentityError("identity JSON is not canonical")
        if set(document) != {field.name for field in fields(cls)} | {
            "format",
            "version",
            "identity_sha256",
        }:
            raise InstallationIdentityError("identity fields do not match the contract")
        if (
            document.get("format") != IDENTITY_FORMAT
            or type(document.get("version")) is not int
            or document["version"] != IDENTITY_VERSION
        ):
            raise InstallationIdentityError("identity format/version is unsupported")
        digest = document.pop("identity_sha256", None)
        if hashlib.sha256(_canonical(document)).hexdigest() != digest:
            raise InstallationIdentityError("identity digest mismatch")
        del document["format"], document["version"]
        try:
            return cls(**document)
        except TypeError as exc:
            raise InstallationIdentityError(
                "identity fields do not match the contract"
            ) from exc

    def require_destination(
        self,
        *,
        environment: str,
        target_database: str,
        candidate_topic: str,
        ordered_topic: str,
    ) -> None:
        for field, expected in (
            ("environment", environment),
            ("target_database", target_database),
            ("candidate_topic", candidate_topic),
            ("ordered_topic", ordered_topic),
        ):
            if getattr(self, field) != expected:
                raise InstallationIdentityError(
                    f"persisted catalog identity conflicts with {field}"
                )


def identity_path(revision_fence_file: str) -> Path:
    fence = Path(revision_fence_file)
    if not fence.is_absolute() or not fence.parent.is_dir():
        raise InstallationIdentityError(
            "identity requires an existing absolute fence directory"
        )
    return fence.parent / IDENTITY_FILENAME


def load_identity(path: Path) -> InstallationIdentity:
    """Read only: missing identity must not generate a new producer on restart."""
    try:
        descriptor = os.open(
            path, os.O_RDONLY | os.O_CLOEXEC | os.O_NOFOLLOW | os.O_NONBLOCK
        )
        with os.fdopen(descriptor, "rb") as source:
            if not stat.S_ISREG(os.fstat(source.fileno()).st_mode):
                raise InstallationIdentityError(
                    "persisted catalog identity is not a regular file"
                )
            raw = source.read(MAX_IDENTITY_BYTES + 1)
    except OSError as exc:
        raise InstallationIdentityError(
            f"cannot read persisted catalog identity: {exc.strerror}"
        ) from exc
    return InstallationIdentity.decode(raw)


def load_or_initialize_identity(
    path: Path,
    *,
    initialize: Callable[[], InstallationIdentity],
) -> InstallationIdentity:
    """Persist once under the shared writer lock, after evidence-based initialization.

    The initializer may inspect an empty DB or validate an existing installation.
    It runs only while holding the lock and only when the file is absent. Any
    validation failure leaves existing state untouched; a corrupt file is never
    replaced. Publication is atomic and fsynced, including the directory entry.
    """
    if not path.is_absolute() or path.name != IDENTITY_FILENAME:
        raise InstallationIdentityError(
            "identity path must use the fixed absolute filename"
        )

    def resolve() -> InstallationIdentity:
        if os.path.lexists(path):
            return load_identity(path)
        identity = initialize()
        if not isinstance(identity, InstallationIdentity):
            raise InstallationIdentityError(
                "identity initializer returned invalid evidence"
            )
        raw = identity.encode()
        descriptor, temporary = tempfile.mkstemp(
            prefix=".catalog-identity-", dir=path.parent
        )
        try:
            with os.fdopen(descriptor, "wb") as target:
                target.write(raw)
                target.flush()
                os.fsync(target.fileno())
            # Never replace an existing identity, including one created by a
            # writer that did not participate in this process's file lock.
            os.link(temporary, path, follow_symlinks=False)
            directory_fd = os.open(path.parent, os.O_RDONLY | os.O_DIRECTORY)
            try:
                os.fsync(directory_fd)
            finally:
                os.close(directory_fd)
        finally:
            os.unlink(temporary)
        return identity

    return FileCatalogMutationSerializer(str(path.parent)).serialize(
        "property-catalog-installation-identity-v1", resolve
    )

File: fi-collector/pkg/propertycatalog/installation_identity.go
package propertycatalog

import (
	"bytes"
	"crypto/sha256"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"regexp"

	"golang.org/x/sys/unix"
)

const (
	InstallationIdentityFilename = "runtime-identity-v1.json"
	installationIdentityFormat   = "futureagi.property-catalog-runtime-identity"
	installationIdentityVersion  = 1
	installationProjectionFormat = "futureagi.property-catalog-values.v1"
	maxInstallationIdentityBytes = 4096
)

// ErrInstallationIdentityMissing means the shared directory exists but its
// descriptor has not been published. Only this error permits startup retry;
// Go must never initialize, repair, or replace the lifecycle-owned identity.
var ErrInstallationIdentityMissing = errors.New("propertycatalog: persisted installation identity is missing")

var installationDatabasePattern = regexp.MustCompile(`^[a-z][a-z0-9_]{0,127}$`)
var installationTopicPattern = regexp.MustCompile(`^[A-Za-z0-9._-]{1,249}$`)

// InstallationIdentity binds an ordered stream to its persisted destination.
// The Python lifecycle initializes it only after inspecting database evidence.
type InstallationIdentity struct {
	Environment       string
	TargetDatabase    string
	CandidateTopic    string
	OrderedTopic      string
	CatalogEpoch      uint16
	ProjectionVersion uint16
	ProducerStreamID  string
	ProjectionFormat  string
}

// Field order is Python json.dumps(sort_keys=True) order. All accepted strings
// are ASCII, so Go's compact JSON encoding equals Python's ensure_ascii=True
// encoding. Re-encoding also rejects duplicate, omitted, null, case-insensitive,
// escaped, or out-of-order fields instead of accepting json.Decoder's coercions.
type installationIdentityDocument struct {
	CandidateTopic    string `json:"candidate_topic"`
	CatalogEpoch      uint16 `json:"catalog_epoch"`
	Environment       string `json:"environment"`
	Format            string `json:"format"`
	IdentitySHA256    string `json:"identity_sha256"`
	OrderedTopic      string `json:"ordered_topic"`
	ProducerStreamID  string `json:"producer_stream_id"`
	ProjectionFormat  string `json:"projection_format"`
	ProjectionVersion uint16 `json:"projection_version"`
	TargetDatabase    string `json:"target_database"`
	Version           uint16 `json:"version"`
}

// InstallationIdentityPath resolves the fixed descriptor beside the revision
// fence. Like the Python identity_path helper, it requires an existing absolute
// fence directory; neither the fence nor the descriptor is created here.
func InstallationIdentityPath(revisionFenceFile string) (string, error) {
	if !filepath.IsAbs(revisionFenceFile) {
		return "", errors.New("propertycatalog: identity requires an absolute revision fence path")
	}
	directory := filepath.Dir(revisionFenceFile)
	info, err := os.Stat(directory)
	if err != nil {
		return "", fmt.Errorf("propertycatalog: inspect installation identity directory: %w", err)
	}
	if !info.IsDir() {
		return "", errors.New("propertycatalog: installation identity parent must be a directory")
	}
	return filepath.Join(directory, InstallationIdentityFilename), nil
}

// LoadInstallationIdentity reads a bounded regular file without following a
// descriptor symlink. O_NONBLOCK ensures a corrupt FIFO cannot hang startup;
// fstat and the bounded read refer to the opened descriptor, avoiding an
// lstat/open race. This function never writes durable state.
func LoadInstallationIdentity(path string) (InstallationIdentity, error) {
	if !filepath.IsAbs(path) || filepath.Base(path) != InstallationIdentityFilename {
		return InstallationIdentity{}, errors.New("propertycatalog: identity path must use the fixed absolute filename")
	}
	if _, err := InstallationIdentityPath(path); err != nil {
		return InstallationIdentity{}, err
	}
	fd, err := unix.Open(path, unix.O_RDONLY|unix.O_CLOEXEC|unix.O_NOFOLLOW|unix.O_NONBLOCK, 0)
	if err != nil {
		readErr := fmt.Errorf("propertycatalog: open installation identity: %w", &os.PathError{Op: "open", Path: path, Err: err})
		if errors.Is(err, os.ErrNotExist) {
			return InstallationIdentity{}, errors.Join(ErrInstallationIdentityMissing, readErr)
		}
		return InstallationIdentity{}, readErr
	}
	source := os.NewFile(uintptr(fd), path)
	defer source.Close()
	info, err := source.Stat()
	if err != nil {
		return InstallationIdentity{}, fmt.Errorf("propertycatalog: inspect installation identity: %w", err)
	}
	if !info.Mode().IsRegular() || info.Size() > maxInstallationIdentityBytes {
		return InstallationIdentity{}, errors.New("propertycatalog: installation identity must be a bounded regular file")
	}
	raw, err := io.ReadAll(io.LimitReader(source, maxInstallationIdentityBytes+1))
	if err != nil {
		return InstallationIdentity{}, fmt.Errorf("propertycatalog: read installation identity: %w", err)
	}
	return ParseInstallationIdentity(raw)
}

// ParseInstallationIdentity accepts exactly the canonical JSON line and SHA256
// contract in Python property_catalog/installation_identity.py. It deliberately
// exposes no identity generator or persistence API.
func ParseInstallationIdentity(raw []byte) (InstallationIdentity, error) {
	if len(raw) == 0 || len(raw) > maxInstallationIdentityBytes || raw[len(raw)-1] != '\n' {
		return InstallationIdentity{}, errors.New("propertycatalog: identity must be a bounded canonical JSON line")
	}
	decoder := json.NewDecoder(bytes.NewReader(raw))
	decoder.DisallowUnknownFields()
	var document installationIdentityDocument
	if err := decoder.Decode(&document); err != nil {
		return InstallationIdentity{}, fmt.Errorf("propertycatalog: decode installation identity: %w", err)
	}
	if err := requireJSONEOF(decoder); err != nil {
		return InstallationIdentity{}, err
	}
	if document.Format != installationIdentityFormat || document.Version != installationIdentityVersion {
		return InstallationIdentity{}, errors.New("propertycatalog: identity format/version is unsupported")
	}
	identity := InstallationIdentity{
		Environment: document.Environment, TargetDatabase: document.TargetDatabase,
		CandidateTopic: document.CandidateTopic, OrderedTopic: document.OrderedTopic,
		CatalogEpoch: document.CatalogEpoch, ProjectionVersion: document.ProjectionVersion,
		ProducerStreamID: document.ProducerStreamID, ProjectionFormat: document.ProjectionFormat,
	}
	if err := identity.validate(); err != nil {
		return InstallationIdentity{}, err
	}
	if !isLowerSHA256(document.IdentitySHA256) {
		return InstallationIdentity{}, errors.New("propertycatalog: identity digest must be a lowercase SHA256")
	}
	canonical, err := json.Marshal(document)
	if err != nil || !bytes.Equal(append(canonical, '\n'), raw) {
		return InstallationIdentity{}, errors.New("propertycatalog: installation identity JSON is not canonical")
	}
	// Maps encode with lexicographically sorted keys, exactly as Python's
	// _canonical(unsigned). The digest field is excluded, not set to empty.
	unsigned, err := json.Marshal(map[string]any{
		"candidate_topic": document.CandidateTopic, "catalog_epoch": document.CatalogEpoch,
		"environment": document.Environment, "format": document.Format,
		"ordered_topic": document.OrderedTopic, "producer_stream_id": document.ProducerStreamID,
		"projection_format": document.ProjectionFormat, "projection_version": document.ProjectionVersion,
		"target_database": document.TargetDatabase, "version": document.Version,
	})
	if err != nil {
		return InstallationIdentity{}, err
	}
	if fmt.Sprintf("%x", sha256.Sum256(unsigned)) != document.IdentitySHA256 {
		return InstallationIdentity{}, errors.New("propertycatalog: installation identity digest mismatch")
	}
	return identity, nil
}

func (identity InstallationIdentity) validate() error {
	if identity.Environment != DevelopmentEnvironment && identity.Environment != ProductionEnvironment {
		return errors.New("propertycatalog: identity environment is unsupported")
	}
	if !installationDatabasePattern.MatchString(identity.TargetDatabase) ||
		identity.TargetDatabase == "default" || identity.TargetDatabase == "system" || identity.TargetDatabase == "information_schema" {
		return errors.New("propertycatalog: identity target database must be an isolated catalog")
	}
	for _, topic := range []string{identity.CandidateTopic, identity.OrderedTopic} {
		if !installationTopicPattern.MatchString(topic) || topic == "." || topic == ".." {
			return errors.New("propertycatalog: identity Kafka topic is invalid")
		}
	}
	if identity.CandidateTopic == identity.OrderedTopic {
		return errors.New("propertycatalog: identity requires distinct Kafka topics")
	}
	if identity.CatalogEpoch == 0 || identity.ProjectionVersion == 0 {
		return errors.New("propertycatalog: identity epoch/projection must be positive UInt16 values")
	}
	if identity.ProjectionFormat != installationProjectionFormat {
		return errors.New("propertycatalog: identity projection format is unsupported")
	}
	return validateCanonicalUUID("identity producer stream", identity.ProducerStreamID)
}

func (identity InstallationIdentity) RequireDestination(environment, targetDatabase, candidateTopic, orderedTopic string) error {
	for _, field := range []struct{ name, actual, expected string }{
		{"environment", identity.Environment, environment},
		{"target_database", identity.TargetDatabase, targetDatabase},
		{"candidate_topic", identity.CandidateTopic, candidateTopic},
		{"ordered_topic", identity.OrderedTopic, orderedTopic},
	} {
		if field.actual != field.expected {
			return fmt.Errorf("propertycatalog: persisted catalog identity conflicts with %s", field.name)
		}
	}
	return nil
}
